### DISCUSSION TITLE
logback 1.3.0 compatibility in 2.1.x

### DISCUSSION THREAD

**Discussion by josephlbarnett** [dropwizard/dropwizard#5821] [Q&A] [answered]
Trying to run with slf4j 2 / logback 1.3.0 in 2.1.1 results in:

```
java.lang.NoClassDefFoundError: ch/qos/logback/classic/jmx/JMXConfigurator
	at io.dropwizard.logging.DefaultLoggingFactory.configure(DefaultLoggingFactory.java:153)
	at io.dropwizard.cli.ConfiguredCommand.run(ConfiguredCommand.java:94)
	at io.dropwizard.cli.Cli.run(Cli.java:78)
```

I see a change in 3.x aimed at fixing this (https://github.com/dropwizard/dropwizard/pull/5781), but can it also be fixed in 2.1.x?

**Comment by zUniQueX** [comment, marked as answer]
Hi @josephlbarnett. We currently don't plan to upgrade Logback to 1.3.0 in the 2.1.x series, maybe we can think of introducing the change in 2.2.0. Logback 1.3.0 dropped support for JMX, so the class cannot be found any more.

If you want to use SLF4J 2.0.0 in your application with Dropwizard 2.1.x, you currently have to provide a custom `LoggingFactory`. Then you'd create a file `src/main/resources/META-INF/services/io.dropwizard.logging.LoggingFactory` in your application and provide the fully qualified class name of your logging factory implementation. Then you can activate the new logging factory with your configuration:
```yaml
logging:
  type: myCustomLoggingFactory
```

**Reply by joschi** [reply]
To reduce the maintenance work (we're currently actively supporting Dropwizard 2.0.x and 2.1.x), I'd vote against doing a Dropwizard 2.2.x branch just for Logback. 😉

**Reply by zUniQueX** [reply]
I agree with that. I just remember some community members wanted to have extended support for Java 8 and maybe they would maintain the 2.x.x series. Then they could introduce support for SLF4J 2 :smile:

**Reply by mattnelson** [reply]
A few options for passively adding this to 2.1.x:

* Support for jackson blackbird was optionally added in 2.1.x by checking at runtime. Could the same be down here? If `JMXConfigurator` is not found, skip that step.
https://github.com/dropwizard/dropwizard/pull/5477
* Add a new configuration to the default logging factory to skip jmx.
```yaml
logging:
  jmx: false
``` 
* Add the custom log factory mentioned above to allow opt-in without having to create it yourself.

**Reply by zUniQueX** [reply]
I'm not sure, if there was a big profit with adding support for SLF4J 2 in Dropwizard 2.1.x. Any reasons, why we would have to care about this upgrade for 2.1.x?

**Reply by mattnelson** [reply]
2.1.x is the latest non-beta release available. slf4j is the defacto logging framework for the entire java ecosystem. As other libraries start moving forward and pulling in slf4j-2.x, those libraries will become incompatible with dropwizard-2.1.x. The justification to add support from this viewpoint would hinge on what the timelines for 2.1.x, 3.x, 4.x look like.

Secondly, and possibly most important. logback-1.2.x is now listed as inactive. If a CVE were to be discovered for that component. The path to resolution would be uplift to 3.x(which is still in beta) or add a custom logging factory, which would likely be identical to `DefaultLoggingFactory` but skipping the jmx bits. I know this is contingent on a "what if" scenario, so this alone should not be the reason to add support. But it can be kept as an option if that situation arises.
https://logback.qos.ch/download.html

**Reply by pstackle** [reply]
I don't think there is a ton of direct benefit to the Dropwizard 2.1.x releases by supporting slf4j 2, but I think it will be very beneficial to unblock users that want to upgrade their own apps so that they can start adopting the new slf4j version from their own code.

I think @mattnelson 's suggestion of gracefully skipping the JMX configuration if `JMXConfigurator` isn't found would be pretty straightforward without introducing too much in the way of a maintenance burden.

According to the Logback 1.3 release notes:
> The ch.qos.logback.classic.jmx package was removed for security reasons and for lack of use.

source: https://logback.qos.ch/news.html

Due to the cited "security reasons" for its removal, I think there is also an argument that Dropwizard 2.1 will be in a better position if users have a path that allows them to reduce security risk by having their app's logging not expose JMX configuration support.

**Reply by zUniQueX** [reply]
@pstackle Makes sense to me. See what you think of #5825.

**Comment by wolpert** [comment]
This is unrelated to the JMX issue... but I'm running into version issues with the request logger for what seems like the same issue... lack of support for newer slf4j/logback. Many of my dependency tree includes libraries that are pulling in slf4j 2.0.6 and logback 1.4.5. When I'm using the latest here's what I get in my logs on every request:

```
00:11:22.880 [dw-admin-38] [trace:] WARN  o.eclipse.jetty.server.HttpChannel - /healthcheck
java.lang.NoClassDefFoundError: org/eclipse/jetty/http/HttpFields$Mutable
	at ch.qos.logback.access.jetty.RequestLogImpl.makeJettyServerAdapter(RequestLogImpl.java:264)
	at ch.qos.logback.access.jetty.RequestLogImpl.log(RequestLogImpl.java:255)
	at org.eclipse.jetty.server.HttpChannel.onCompleted(HttpChannel.java:911)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:467)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.ClassNotFoundException: org.eclipse.jetty.http.HttpFields$Mutable
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	... 16 common frames omitted
```

I'm not sure what the right solution is... but going backwards to SLF4J 1.x and logback 1.2 doesn't seem possible

**Reply by zUniQueX** [reply]
Hi @wolpert. What you're experiencing here is a bug in `logback-access`, which lacks full support for Jetty 10.x. I created a PR for that some time ago that's still unmerged.

In the `release/3.0.x` branch we provide a workaround for this issue in the `dropwizard-request-logging` module.

**Comment by wolpert** [comment]
Well... I'm using 2.1.4 version of dropwizard. I'm assuming I cannot pull in the fix wholesale to my environment. I'll peak at the branch and see what I can pilfer. 

So... when is the next 3.x or 4.x release :D

**Reply by joschi** [reply]
Unless any regressions are reported, Dropwizard [3.0.0-beta.5](https://github.com/dropwizard/dropwizard/discussions/6736) and [4.0.0-beta.4](https://github.com/dropwizard/dropwizard/discussions/6737) will be published as their respective GA release.

So give them a try and report anything not working for you. 😉

**Reply by wolpert** [reply]
I'll upgrade github.com/wolpert/svarm today to 4.0.0-beta.4 and will give details on the upgrade.

**Reply by wolpert** [reply]
*tl;dr* +1 It worked for me.

*Details*:
Upgrade from dropwizard 2.x to 4.x for a moderately complicated project took about 1 hour with full testing working. Steps I did:

1. Updated dependency to dropwizard 4.0.0-beta.4
2. Changed the javax-> jarkartax namespace project wise
3. Included open-feign's jakarta contract instead of the jaxrs2
4. Removed wacky dependency stuff I was doing for logback upgrade (manually pulling in javax.servlet spec) Finding this was my biggest timesink in the process
5. Fixed imports of dropwizard. (Environment is in io.dropwizard.core.setup.Environment for example)

I would say finding the dependency issues that I've added to the build to make dropwizard work with logback or
whatnot was the biggest issue I ran into. I haven't done any performance testing yet, but my end-to-end test did work fine. (Involves setup of a docker images with multiple servers. Hitting one server will end up calling others. OpenFeign used for all server interactions.)

If you want to see the commit for the upgrade, it is [here](https://github.com/wolpert/svarm/commit/0d960e03db4c615a715d218206acf23603639afd)

**Reply by joschi** [reply]
@wolpert Thanks for the detailed description of the upgrade process!

> Changed the javax-> jarkartax namespace project wise

I think this could even be (semi-) automated with OpenRewrite and https://docs.openrewrite.org/reference/recipes/java/migrate/jakarta.

**Reply by wolpert** [reply]
I use intelij so it was fairly easy to do it across the project. I've never seen openrewrite before... gonna review.
